primitives/npos-elections/src/phragmms.rs

```diff
@@ -0,0 +1,380 @@
+	initial_candidates: Vec<AccountId>,
+	initial_voters: Vec<(AccountId, VoteWeight, Vec<AccountId>)>,
+) -> Result<ElectionResult<AccountId, P>, &'static str>
+	where ExtendedBalance: From<InnerOf<P>>
+{
+	let (candidates, mut voters) = setup_inputs(initial_candidates, initial_voters);
+
+	let mut winners = vec![];
+	for round in 0..to_elect {
+		let round_winner = calculate_max_score::<AccountId, P>(&candidates, &voters);
+		apply_elected::<AccountId>(&mut voters, Rc::clone(&round_winner));
+
+		round_winner.borrow_mut().round = round;
+		round_winner.borrow_mut().elected = true;
+		winners.push(round_winner);
+
+		balance(&mut voters, 2, 0);
+	}
+
+	let mut assignments = voters.into_iter().filter_map(|v| v.into_assignment()).collect::<Vec<_>>();
+	let _ = assignments.iter_mut().map(|a| a.try_normalize()).collect::<Result<(), _>>()?;
+	let winners = winners.into_iter().map(|w_ptr|
+		(w_ptr.borrow().who.clone(), w_ptr.borrow().backed_stake)
+	).collect();
+
+	Ok(ElectionResult { winners, assignments })
+}
+
+/// Find the candidate that can yield the maximum score for this round.
+///
+/// Returns a new `CandidatePtr` to the winner candidate. The score of the candidate is updated and
+/// can be read from the returned pointer.
+///
+/// This is an internal part of the [`balanced_heuristic`].
+pub(crate) fn calculate_max_score<AccountId: IdentifierT, P: PerThing>(
+	candidates: &[CandidatePtr<AccountId>],
+	voters: &[Voter<AccountId>],
+) -> CandidatePtr<AccountId> where ExtendedBalance: From<InnerOf<P>> {
+	for c_ptr in candidates.iter() {
+		let mut candidate = c_ptr.borrow_mut();
+		if !candidate.elected {
+			candidate.score = Rational128::from(1, P::ACCURACY.into());
+		}
+	}
+
+	for voter in voters.iter() {
+		let mut denominator_contribution: ExtendedBalance = 0;
+
+		// gather contribution from all elected edges.
+		for edge in voter.edges.iter() {
+			let edge_candidate = edge.candidate.borrow();
+			if edge_candidate.elected {
+				let edge_contribution: ExtendedBalance = P::from_rational_approximation(
+					edge.weight,
+					edge_candidate.backed_stake,
+				).deconstruct().into();
+				denominator_contribution += edge_contribution;
+			}
+		}
+
+		// distribute to all _unelected_ edges.
+		for edge in voter.edges.iter() {
+			let mut edge_candidate = edge.candidate.borrow_mut();
+			if !edge_candidate.elected {
+				let prev_d = edge_candidate.score.d();
+				edge_candidate.score = Rational128::from(1, denominator_contribution + prev_d);
+			}
+		}
+	}
+
+	// finalise the score value, and find the best.
+	let mut best_score = Rational128::zero();
+	let mut best_candidate = Rc::clone(&candidates[0]);
+	for c_ptr in candidates.iter() {
+		let mut candidate = c_ptr.borrow_mut();
+		if candidate.approval_stake > 0  {
+			// finalise the score value.
+			let score_d = candidate.score.d();
+			let one: ExtendedBalance = P::ACCURACY.into();
+			let score_n = candidate.approval_stake.checked_mul(one).unwrap_or_else(|| Bounded::max_value());
```

Comment:
so score_n must fit in u128, this means that maximum approval_stake allowed is `u128::max_value()/ACCURACY` so if user choose a too big accuracy then it can saturate, maybe it worth put into phragmms doc, so that user don't try to do use to high accuracy

(as a note if accuracy is PerQuintill, max_approval_stake is `(2^128 - 1 )* 10^-18 * 10^-12 = 350_000_000 KSM`)

Reply:
Actually I am thinking of using big-unit directly here to allow higher approval_stakes. I was debugging this and this actually does saturate now in kusama, which is bad, and which is one of the reasons that I have not yet used in this implementation anywhere. This is one of the details that need to be sorted.

Reply:
I wrote an extensive inline doc about this. With Perbill, it is pretty much fine because the same max that you calculated above would be: 

```
>>> '{:,}'.format((2 ** 128 - 1) / 10**9 / 10**12)
'340,282,366,920,938,463'
```


Given that we want to use this offchain, the `P` will actually be `PerU16` which will give: 

```
>>> '{:,}'.format((2 ** 128 - 1) / 2**16 / 10**12)
'5,192,296,858,534,827,628,530'
```

So for now I would keep it as it is.

Reply:
For the near future I would create this: 

```
trait UnwrapOrElseWithLog {
  fn unwrap_or_else_with_log() {
     // log that an overflow happened somewhere and and saturated it
    Bounded::max_value() // or Bounded::min_value()
  }
}
```

And replace it in all the _dangerous_ places that we recklessly do `unwrap_or_else(max_value)` or `saturating_xxx`.

This should inform us in the future if we hit the numeric boundaries of u128. Frankly, with Kusama/Polkadot's token type I think this will never happen. This is more for the sake of long term future and chains that would have a custom funky balance type.

Reply:
yes ok